docs/runtime.adoc

```diff
@@ -0,0 +1,94 @@
+
+* `THREAD_POOL` - process one partition with the platform threads of count `decaton.partition.concurrency`
+* `VIRTUAL_THREAD` - process each different key on a different virtual thread. `decaton.partition.concurrency` configuration is meaningless with this runtime. **Requires JDK21 or higher**
+
+== Usage
+
+Just set `SubPartitionRuntime` by calling `SubscriptionBuilder#subPartitionRuntime()`. By default it is set to `SubPartitionRuntime.THREAD_POOL`.
+
+[source,java]
+----
+SubscriptionBuilder.newBuilder("my-subscription")
+        .subPartitionRuntime(SubPartitionRuntime.VIRTUAL_THREAD)
+...
+        .build();
+----
+
+IMPORTANT: When using the `VIRTUAL_THREAD`, be extremely careful for the use of `synchronized` primitive in your/library code that might be executed in `DecatonProcessor#process()` method, as it results "pinning" the virtual thread and could eliminate performance gain of using virtual thread completely.
+Visit https://docs.oracle.com/en/java/javase/21/core/virtual-threads.html#GUID-704A716D-0662-4BC7-8C7F-66EE74B1EDAD[Scheduling Virtual Threads and Pinned Virtual Threads] for the detail.
+
+== Performance
+
+As for the reference of performance characteristic of each runtime, here's the benchmarking result.
+
+The machine spec used for the benchmark is as follows.
+
+* CPU: Intel(R) Xeon(R) 2.20GHz, 8 core
+* Memory: 16GiB
+
+The benchmark was run using the https://github.com/line/decaton/tree/master/benchmark[benchmark module] with the following command line options.
+
+[source]
+----
+./debm.sh \
+  --title "Decaton" \
+  --runner com.linecorp.decaton.benchmark.DecatonRunner \
+  --runs 3 \
+  --format json \
+  --tasks 10000 \
+  --warmup 100000 \
+  --simulate-latency=$latency \
+  --latency-count=5 \
+  --param=decaton.partition.concurrency=300 \
+  --param=decaton.subpartition.runtime={THREAD_POOL|VIRTUAL_THREAD}
+----
+
+image::runtime-benchmark-result.png[]
+
+We can read from the result that the `VIRTUAL_THREAD` runtime provides much stable throughput around 23k tasks/sec regardless to the amount of duration taken for I/O in processing a task.
+
+
+== Appendix
+
+=== `decaton.partition.concurrency` tuning
+
+As written previously in this document, having too much platform (OS) threads causes significant overhead and throughput growth is capped at some point.
+To learn where is that point, a benchmark was run to determine at which `decaton.partition.concurrency` value do we hit throughput peak.
+
+The following command is used with increasing concurrency value. The machine spec is as same as <<Performance>>.
+
+[source]
+----
+./debm.sh \
+  --title "Decaton" \
+  --runner com.linecorp.decaton.benchmark.DecatonRunner \
+  --runs 2 \
+  --format json \
+  --tasks 10000 \
+  --warmup 100000 \
+  --simulate-latency=4 \
+  --latency-count=5 \
+  --param=decaton.partition.concurrency=$conc \
+  --param=decaton.subpartition.runtime=THREAD_POOL
+----
+
+image::runtime-benchmark-conc-peak.png[]
```

Comment:
Though I understand that the focus of this appendix is about throughput-cap, I guess this graph might make confusing impression because 30000 with 20ms (4ms * 5) latency is still higher than vthread throughput shown in above graph. (So people may think "Got it, threadpool-runtime throughput will be capped by 30000 in 20ms-latency scenario. But vthread-runtime can't achieve 30000 in the first place....")

Isn't it better to show throughput cap with more high simulated latency like 100ms?

Reply:
Hm, changed my mind.
Maybe I'm worrying too much. LGTM